[EDU 6340] Refactor: azion iac doc (#1560)
aziontech/docs: src/content/docs/en/pages/devtools/cli/azion-cli/config/azion-config-js.mdx

- Rules Engine
- Network List
- Purge Settings
- Firewall Settings
- WAF Settings

:::note
The default properties and values generated in the project's initial build are tested and work properly. In case this file is altered, we no longer guarantee it works properly, so use it wisely.
:::

> ---

Reviewer: ```suggestion
---

```

## API Reference

### `defineConfig`

Configures and validates the options for the Azion Edge Application.

**Parameters**:

- `config: AzionConfig` - The configuration object for the Azion Edge Application.

### `processConfig`

Processes the configuration object and returns a manifest.

### `convertJsonConfigToObject`

> Converts a Azion JSON configuration object to a AzionConfig object.

Reviewer: ```suggestion
Converts an Azion JSON configuration object to an AzionConfig object.
```

**Parameters**:

- `config: string` - The JSON configuration object.

**Parameters**:

- `config: AzionConfig` - The configuration object for the Azion Edge Application.

---

> ## Types

Reviewer: ```suggestion

## Types
```

> ### `AzionConfig`

Reviewer: não tem description como nos outros types?

**Properties**:

- `build?: AzionBuild` - The build configuration.
- `domain?: AzionDomain` - The domain object.
- `origin?: AzionOrigin[]` - List of origins.
- `cache?: AzionCache[]` - List of cache settings.
- `rules?: AzionRules[]` - List of edge rules.
- `purge?: AzionPurge[]` - List of URLs or CacheKeys to purge.
- `networkLists?: AzionNetworkList[]` - List of network lists.
- `waf?: AzionWaf[]` - List of WAF configurations.

### `AzionBuild`

Type definition for the build configuration.

**Properties**:

- `builder?: 'esbuild' | 'webpack'` - The builder to use.
- `preset?: { name: string; }` - The preset to use.
- `entry?: string` - The entry file.
- `polyfills?: boolean` - Whether to include polyfills.
- `worker?: boolean` - Whether to build a owner worker.
- `custom?: Record<string, any>` - Custom build configuration.
- `memoryFS?: { injectionDirs: string[], removePathPrefix: string }` - Memory file system configuration.

### `AzionDomain`

Type definition for the domain configuration.

**Properties**:

- `name: string` - The domain name.
- `cnameAccessOnly?: boolean` - Whether to restrict access only to CNAMEs.
- `cnames?: string[]` - List of CNAMEs for the domain.

> - `id?: number` - ID of the domain.
> - `edgeApplicationId?: number` - ID of the edge application.
> - `edgeFirewallId?: number` - ID of the edge firewall.
> - `digitalCertificateId?: string | number | null` - ID of the digital certificate.

Reviewer: ```suggestion
- `id?: number` - Domain ID.
- `edgeApplicationId?: number` - Edge application ID.
- `edgeFirewallId?: number` - Edge firewall ID.
- `digitalCertificateId?: string | number | null` - Digital certificate ID.
```

- `active?: boolean` - Whether the domain is active.
- `mtls?: MTLSConfig` - Configuration for mTLS.
  - `verification: 'enforce' | 'permissive'` - mTLS verification mode.

>   - `trustedCaCertificateId: number` - ID of the trusted CA certificate.

Reviewer: ```suggestion
  - `trustedCaCertificateId: number` - Trusted CA certificate ID.
```

  - `crlList?: number[]` - List of Certificate Revocation Lists (CRLs).

### `AzionOrigin`

Type definition for the origin configuration.

**Properties**:

> - `id?: number` - ID of the origin.
> - `key?: string` - Key for the origin.
> - `name: string` - Name of the origin.
> - `type: string` - Type of the origin (For exemple, 'single_origin', 'load_balancer').

Reviewer: ```suggestion
- `id?: number` - Origin ID.
- `key?: string` - Origin key.
- `name: string` - Origin name.
- `type: string` - Type of the origin. For example: 'single_origin', 'load_balancer'.
```

### `AzionResponseRule`

Type definition for the response rule configuration.

**Properties**:

- `name: string` - Name of the response rule.
- `description?: string` - Description of the response rule.
- `active?: boolean` - Whether the rule is active.
- `match: string` - Match criteria for the rule.
- `variable?: string` - Variable to be used in the match.
- `behavior?: ResponseBehavior` - Behavior to apply when the rule matches.

  - `setCookie?: string | null` - Set a cookie.
  - `setHeaders?: string[]` - Set headers.
  - `deliver?: boolean | null` - Deliver the content.
  - `capture?: { match: string; captured: string; subject: string }` - Capture configuration.
  - `enableGZIP?: boolean | null` - Enable GZIP compression.
  - `filterCookie?: string | null` - Filter a cookie.
  - `filterHeader?: string | null` - Filter a header.
  - `runFunction?: { path: string; name?: string | null }` - Run a serverless function.
  - `redirectTo301?: string | null` - Redirect with 301 status.
  - `redirectTo302?: string | null` - Redirect with 302 status.

### `AzionRules`

Type definition for the rule set.

**Properties**:

> - `request: AzionRequestRule[]` - Ruleset for Request phase.
> - `response?: AzionResponseRule[]` - Ruleset for Response phase.

Reviewer: ```suggestion
- `request: AzionRequestRule[]` - Rule set for the Request phase.
- `response?: AzionResponseRule[]` - Rule set for the Response phase.
```

### `AzionPurge`

Type definition for the purge configuration.

**Properties**:

- `type: 'url' | 'cachekey' | 'wildcard'` - The type of purge to be performed.
- `urls: string[]` - List of URLs or patterns to be purged.
- `method?: 'delete'` - HTTP method for the purge request.
- `layer?: 'edge_caching' | 'l2_caching'` - Cache layer to be purged.

### `AzionNetworkList`

Type definition for the network list configuration.

**Properties**:

> - `id: number` - ID of the network list.
> - `listType: 'ip_cidr' | 'asn' | 'countries'` - Type of the network list.

Reviewer: ```suggestion
- `id: number` - Network list ID.
- `listType: 'ip_cidr' | 'asn' | 'countries'` - Network list type.
```

- `name: string` - Name of the firewall.
- `domains?: string[]` - List of domains associated with the firewall.
- `active?: boolean` - Whether the firewall is active.
- `edgeFunctions?: boolean` - Whether Edge Functions are enabled.
- `networkProtection?: boolean` - Whether Network Protection is enabled.
- `waf?: boolean` - Whether WAF is enabled.
- `variable?: RuleVariable` - Variable to be used in matches.
- `rules?: AzionFirewallRule[]` - List of firewall rules.
- `debugRules?: boolean` - Whether debug mode is enabled for rules.

### `AzionFirewallRule`

Type definition for firewall rules.

**Properties**:

- `name: string` - Name of the rule.
- `description?: string` - Description of the rule.
- `active?: boolean` - Whether the rule is active.
- `match?: string` - Match criteria for the rule.
- `variable?: RuleVariable` - Variable to be used in the match.
- `criteria?: AzionFirewallCriteria[]` - Array of criteria for complex conditions.
- `behavior: AzionFirewallBehavior` - Behavior to be applied when the rule matches.

### `AzionFirewallBehavior`

Type definition for firewall rule behaviors.

**Properties**:

- `runFunction?: { path: string }` - Run a serverless function.

> - `setWafRuleset?: { wafMode: FirewallWafMode; wafId: string }` - Set WAF ruleset.

Reviewer: ```suggestion
- `setWafRuleset?: { wafMode: FirewallWafMode; wafId: string }` - Set WAF rule set.
```

- `setRateLimit?: {` - Set rate limit configuration.
    - `type: FirewallRateLimitType` - Rate limit type (second, minute, hour).
    - `limitBy: FirewallRateLimitBy` - Rate limit by (client_ip, global, token).
    - `averageRateLimit: string` - Average rate limit.
    - `maximumBurstSize: string` - Maximum burst size.
  - `deny?: boolean` - Deny the request.
  - `drop?: boolean` - Drop the request.
  - `setCustomResponse?: {` - Set custom response.
    - `statusCode: number | string` - HTTP status code (200-499).
    - `contentType: string` - Response content type.
    - `contentBody: string` - Response content body.

### `AzionFirewallCriteria`

Type definition for firewall rule criteria.

**Properties**:

  - `variable: RuleVariable` - Variable to be evaluated.
  - `conditional: RuleConditional` - Conditional type.
  - `operator: RuleOperatorWithValue | RuleOperatorWithoutValue` - Comparison operator.
  - `inputValue?: string` - Input value for comparison (required for operators with value).

### `AzionWaf`

Type definition for the Web Application Firewall (WAF) configuration.

**Properties**:

>   - `id?: number` - ID of the WAF.
>   - `name: string` - Name of the WAF.

Reviewer: ```suggestion
  - `id?: number` - WAF ID.
  - `name: string` - WAF name.
```

  - `bypassCache?: boolean | null` - Bypass cache.
  - `httpToHttps?: boolean | null` - Force HTTPS.
  - `redirectTo301?: string | null` - Redirect with 301 status.
  - `redirectTo302?: string | null` - Redirect with 302 status.
  - `forwardCookies?: boolean | null` - Forward cookies.
  - `setCookie?: string | null` - Set a cookie.
  - `deliver?: boolean | null` - Deliver the content.
  - `capture?: { match: string; captured: string; subject: string }` - Capture configuration.
  - `runFunction?: { path: string; name?: string | null }` - Run a serverless function.
  - `setCache?: string | { name: string; browser_cache_settings_maximum_ttl?: number | null; cdn_cache_settings_maximum_ttl?: number | null }` - Set cache configuration.

### `AzionResponseRule`

Type definition for the response rule configuration.

**Properties**:

- `name: string` - Name of the response rule.
- `description?: string` - Description of the response rule.
- `active?: boolean` - Whether the rule is active.
- `match: string` - Match criteria for the rule.
- `variable?: string` - Variable to be used in the match.
- `behavior?: ResponseBehavior` - Behavior to apply when the rule matches.

> 

Reviewer 2: ```suggestion
```

- `description?: string` - Description of the rule.
- `active?: boolean` - Whether the rule is active.
- `match?: string` - Match criteria for the rule.
- `variable?: RuleVariable` - Variable to be used in the match.
- `criteria?: AzionFirewallCriteria[]` - Array of criteria for complex conditions.
- `behavior: AzionFirewallBehavior` - Behavior to be applied when the rule matches.

### `AzionFirewallBehavior`

Type definition for firewall rule behaviors.

**Properties**:

- `runFunction?: { path: string }` - Run a serverless function.
- `setWafRuleset?: { wafMode: FirewallWafMode; wafId: string }` - Set WAF rule set.
- `setRateLimit?: {` - Set rate limit configuration.
    - `type: FirewallRateLimitType` - Rate limit type (second, minute, hour).
    - `limitBy: FirewallRateLimitBy` - Rate limit by (client_ip, global, token).
    - `averageRateLimit: string` - Average rate limit.
    - `maximumBurstSize: string` - Maximum burst size.
  - `deny?: boolean` - Deny the request.
  - `drop?: boolean` - Drop the request.
  - `setCustomResponse?: {` - Set custom response.
    - `statusCode: number | string` - HTTP status code (200-499).
    - `contentType: string` - Response content type.

>     - `contentBody: string` - Response content body.

Reviewer 2: Is the indentation here correct? Also, on lines 274 and 281, the names of properties are ending in `{`

### `AzionFirewallCriteria`

Type definition for firewall rule criteria.

**Properties**:

  - `variable: RuleVariable` - Variable to be evaluated.
  - `conditional: RuleConditional` - Conditional type.
  - `operator: RuleOperatorWithValue | RuleOperatorWithoutValue` - Comparison operator.
  - `inputValue?: string` - Input value for comparison (required for operators with value).

### `AzionWaf`

Type definition for the Web Application Firewall (WAF) configuration.

**Properties**:

  - `id?: number` - WAF ID.
  - `name: string` - WAF name.
  - `active: boolean` - Whether the WAF is active.

>   - `mode: WafMode` - WAF mode (learning, blocking and counting).

Reviewer 3: ```suggestion
  - `mode: WafMode` - WAF mode (learning, blocking, and counting).
```
